impeller/compiler/shader_lib/impeller/gradient.glsl

```diff
@@ -7,27 +7,124 @@
 
 #include <impeller/texture.glsl>
 
+mat3 IPMapToUnitX(vec2 p0, vec2 p1) {
+  // Returns a matrix that maps [p0, p1] to [(0, 0), (1, 0)]. Results are
+  // undefined if p0 = p1.
+  return mat3(0.0, -1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0) *
+         inverse(mat3(p1.y - p0.y, p0.x - p1.x, 0.0, p1.x - p0.x, p1.y - p0.y,
+                      0.0, p0.x, p0.y, 1.0));
+}
+
 /// Compute the t value for a conical gradient at point `p` between the 2
-/// circles defined by (c0, r0) and (c1, r1).
-///
-/// This assumes that c0 != c1.
-float IPComputeConicalT(vec2 c0, float r0, vec2 c1, float r1, vec2 p) {
-  float w = 1.0;
-  float result = 0.0;
-  vec2 ab = c1 - c0;
-  float dr = r1 - r0;
-  // Set sample rate to a minimum for the case where c0 and c1 are close.
-  float delta = 1.0 / max(length(ab), 100.0);
-  while (w >= 0.0) {
-    vec2 cw = w * ab + c0;
-    float rw = w * dr + r0;
-    if (length(p - cw) <= rw) {
-      result = w;
-      break;
+/// circles defined by (c0, r0) and (c1, r1). The returned vec2 encapsulates 't'
+/// as its x component and validity status as its y component, with positive y
+/// indicating a valid result.
+vec2 IPComputeConicalT(vec2 c0, float r0, vec2 c1, float r1, vec2 pos) {
```

Comment:
You should add a comment that points to the graphite implementation.  If there are upstream bugfixes we'll want to know where to grab them.